Email assistant webapp (#126)
DAGWorks-Inc/burr: examples/web-server/README.md

# Running Burr in a web server

Burr is meant to run interactive apps. This means running it as part of a web-service that

> responds to requests, maanges state, and documents its capabilities. The interactive nature of Burr

Reviewer: ```suggestion
responds to requests, manages state, and documents its capabilities. The interactive nature of Burr
```

(moving in/out of programmatic control) means we want to think carefully about how to expose

> our burr applications to the web.

Reviewer: ```suggestion
our Burr applications to the web.
```

### Additional concerns

#### Scaling

But [is this webscale](https://www.youtube.com/watch?v=b2F-DItXtZs)? As anything, it depends on how you implement it.
Two factors determine the scalability of this system:

1. database layer -- can the database support the volume of inputs/outputs?
2. compute layer -- can the server run fast enough to keep up with the users?

For the database layer, it depends largely on the underlying database, as well as the
schema you use. That said, Burr makes it easier due to natural partitioning of the data
into `application_id` and `partition_key` (the latter could be the `user ID`), allowing common
operations (such as _give me all applications for X user_ and _give me the state of application Y_)
simple if you index your state table on the application ID and `partition_key`.

For the compute layer, you can simply scale horizontally. The only tricky aspect is ensuring state synchronization
and locking. As we cached the application object, we could potentially get into a position
in which the state is out of sync. To solve this, you can either:

> 1. Use a locking method (either in the database) to ensure that only one server is running a given application at any point
> 2. Use sticky sessions/sharding to ensure that a given user always hits the same server
> 
> Or possibly both.

Reviewer: ```suggestion
1. Use a locking method (either in the database) to ensure that only one server is running a given application at any point
2. Use sticky sessions/sharding to ensure that a given user always hits the same server
3. Handle forking of state at the persistence layer with a custom implementation

Or possibly some combination of the above.
```

Author: Yeah thought about it but it feels messy/scary. Let's add it in and see what people use?

#### Async

While we implemented synchronous calls, you can easily make these async by using `async def` and `await` in the appropriate places,
and using the `arun` method in Burr. Read more about async capabilities in [applications](https://burr.dagworks.io/concepts/state-machine/),
and [actions](https://burr.dagworks.io/concepts/actions/).

#### Streaming

You can use streaming to send back the stream of the output at any given point. You do this by creating a
[streaming action](https://burr.dagworks.io/concepts/streaming-actions/). You can then integrate with the
streaming respose in FastAPI to send back the stream of the output. You can do this with any steps
(intermediate or final) in your application.

#### Authentication/Data access

While Burr does not operate at the data access layer, this can be easily handles at the  application layer.
Any authentication system will tell you the user ID, which you can look in your DB to determine access
to your partition key.

> ## Wrapup

Reviewer: ```suggestion
## Wrap up
```
